### PR TITLE
DAOS-2113 control: Move nvme/scm providers out of server

### PR DESCRIPTION
**Pull request by mjmac**
NB: I think I'm going to go in a different direction. There isn't a lot of benefit to landing this as-is and then pulling all of the RPC goop back out of the storage package. Going to leave this PR for now, though.

Moves the NVMe/SCM storage provider code into the storage package
as a step toward getting SPDK out of daos_server.

By itself, this commit doesn't really change very much, but it's a step along the way.
There is still a lot of tight coupling between RPC stuff and core storage logic that
I'm working on pulling apart in subsequent PRs.

### CODE REVIEW

File: src/control/server/storage/scm.go [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+	return s.prep.PrepReset(state)
+}
+
+// Discover method implementation for ScmProvider
+func (s *ScmProvider) Discover() error {
+	if s.initialized {
+		return nil
+	}
+
+	mms, err := s.ipmctl.Discover()
```
Comment (tanabarr) [inline src/control/server/storage/scm.go]: please make sure the changes from https://github.com/daos-stack/daos/pull/1079 get carried over into this PR if it lands before

File: src/control/server/storage/scm_test.go [hunk 2]
```diff
@@ -1,4 +1,48 @@
 //
+// (C) Copyright 2019 Intel Corporation.
```
Comment (tanabarr) [inline src/control/server/storage/scm_test.go]: seems to have multiple copyright entries on this file

File: src/control/server/ctl_storage_rpc.go [hunk 3]
```diff
@@ -60,15 +45,15 @@ func (c *StorageControlService) doNvmePrepare(req *pb.PrepareNvmeReq) (resp *pb.
 	})
 
 	if err != nil {
-		resp.State = newState(c.log, pb.ResponseStatus_CTRL_ERR_NVME, err.Error(), "", msg)
+		resp.State = storage.NewState(c.log, pb.ResponseStatus_CTRL_ERR_NVME, err.Error(), "", msg)
```
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/ctl_storage_rpc.go]: This was a rough first pass with the primary goal of getting everything storage related into the storage package and making the tests pass with minimal modifications. Next step is to take a look at things and figure out the best way to cleanly decouple different areas of concern. This exercise has exposed quite a bit of tight coupling between the RPC handlers and the core storage provider code, so the goal will be to pull any RPC-specific stuff out of the storage package.

File: src/control/server/storage/scm_test.go (hunk 2 above)
Comment (mjmac) [reply src/control/server/storage/scm_test.go]: Weird, not sure how that happened.

File: src/control/server/ctl_storage_rpc.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mjmac) [reply src/control/server/ctl_storage_rpc.go]: I'm thinking that this pass shouldn't be concerned with making a clean separation between core/RPC, actually. It's already a big PR, and keeping the steps separate should make it easier to review.

File: src/control/server/storage/scm.go (hunk 1 above)
Comment (mjmac) [reply src/control/server/storage/scm.go]: done

File: src/control/server/storage/scm_test.go (hunk 2 above)
Comment (mjmac) [reply src/control/server/storage/scm_test.go]: fixed

**Comment by mjmac** [comment]
Going to close this, as I have a different approach that I think will work a lot better.

**Pull request closed without merging**